### PR TITLE
[SPARK-21092][SQL] Wire SQLConf in logical plan and expressions

### PR DESCRIPTION
**Pull request by rxin**
## What changes were proposed in this pull request?
It is really painful to not have configs in logical plan and expressions. We had to add all sorts of hacks (e.g. pass SQLConf explicitly in functions). This patch exposes SQLConf in logical plan, using a thread local variable and a getter closure that's set once there is an active SparkSession.

The implementation is a bit of a hack, since we didn't anticipate this need in the beginning (config was only exposed in physical plan). The implementation is described in `SQLConf.get`.

In terms of future work, we should follow up to clean up CBO (remove the need for passing in config).

## How was this patch tested?
Updated relevant tests for constraint propagation.

### CODE REVIEW

**Comment by rxin** [comment]
Note that this patch is based on https://github.com/apache/spark/pull/18298. Once we merge that one the diff will become smaller.

**Comment by rxin** [comment]
cc @wzhfy

**Comment by wzhfy** [comment]
Since we have a thread local variable in `SQLConf`, can we just provide getter/setter methods, get it in `QueryPlan` and set it in `SparkSession`?

**Comment by rxin** [comment]
The issue is that SparkSession might change the way they are wired and it's not always the case that when we create a new thread, we set the thread local conf.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlanConstraints.scala [hunk 1]
```diff
@@ -27,18 +27,20 @@ trait QueryPlanConstraints[PlanType <: QueryPlan[PlanType]] { self: QueryPlan[Pl
...
-      constraints
+  lazy val constraints: ExpressionSet = {
+    if (conf.constraintPropagationEnabled) {
+      ExpressionSet(
+        validConstraints
+          .union(inferAdditionalConstraints(validConstraints))
+          .union(constructIsNotNullConstraints(validConstraints))
+          .filter { c =>
+            c.references.nonEmpty && c.references.subsetOf(outputSet) && c.deterministic
+          }
```
Comment (sameeragarwal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlanConstraints.scala]: should we also bound their size as you suggested?

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlanConstraints.scala]: That belongs in a separate patch.

**Comment by wzhfy** [comment]
Got it, thanks.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/ConstraintPropagationSuite.scala [hunk 2]
```diff
@@ -402,17 +403,19 @@ class ConstraintPropagationSuite extends SparkFunSuite {
...
       .groupBy('a, 'c, 'b)('a, 'c.as("c1"), count('a).as("a3")).select('c1, 'a, 'a3)
 
-    verifyConstraints(aliasedRelation.analyze.getConstraints(constraintPropagationEnabled = true),
-      aliasedRelation.analyze.constraints)
-    assert(aliasedRelation.analyze.getConstraints(constraintPropagationEnabled = false).isEmpty)
+    SQLConf.get.setConf(SQLConf.CONSTRAINT_PROPAGATION_ENABLED, true)
+    assert(aliasedRelation.analyze.constraints.nonEmpty)
+
+    SQLConf.get.setConf(SQLConf.CONSTRAINT_PROPAGATION_ENABLED, false)
+    assert(aliasedRelation.analyze.constraints.isEmpty)
```
Comment (wzhfy) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/ConstraintPropagationSuite.scala]: unset the config in the end

**Comment by wzhfy** [comment]
LGTM except one minor comment.

**Comment by rxin** [comment]
Merging in master.

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 3]
```diff
@@ -87,6 +87,11 @@ class SparkSession private(
 
   sparkContext.assertNotStopped()
 
+  // If there is no active SparkSession, uses the default SQL conf. Otherwise, use the session's.
+  SQLConf.setSQLConfGetter(() => {
+    SparkSession.getActiveSession.map(_.sessionState.conf).getOrElse(SQLConf.getFallbackConf)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]: shall we move this to `SparkSession.Builder.getOrCreate`?
